orangecontrib/spectroscopy/data.py

```diff
@@ -1014,16 +1014,148 @@ def read_v2(self):
+            file_gsf_p = self.filename
+            file_gsf_a = self.filename.replace('P raw.gsf', 'A raw.gsf')
+            file_html = folder_file + '.html'
+        elif 'A' in file_channel:
+            self.channel_a = file_channel
+            self.channel_p = file_channel.replace('A', 'P')
+            file_gsf_a = self.filename
+            file_gsf_p = self.filename.replace('A raw.gsf', 'P raw.gsf')
+            file_html = folder_file + '.html'
+
+
+        data_gsf_a = self._gsf_reader(file_gsf_a)
+        data_gsf_p = self._gsf_reader(file_gsf_p)
+        info = self._html_reader(file_html)
+
+        final_data, parameters, final_metas = self._format_file(data_gsf_a, data_gsf_p, info)
+
+
+        metas = [Orange.data.ContinuousVariable.make("run"),
+                Orange.data.ContinuousVariable.make("row"),
+                Orange.data.ContinuousVariable.make("column"),
+                Orange.data.StringVariable.make("channel")]
+
+        domain = Orange.data.Domain([], None, metas=metas)
+        meta_data = Table.from_numpy(domain, X=np.zeros((len(final_data), 0)),
+                                     metas=np.asarray(final_metas, dtype=object))
+        
+        meta_data.attributes = parameters
+
+        depth = np.arange(0, int(parameters['Pixel Area (X, Y, Z)'][3]))
+
+        return depth, final_data, meta_data
+
+    def _format_file(self, gsf_a, gsf_p, parameters):
+        
+        info = {}
+        for row in range(len(parameters)):
+            info.update({parameters[row,0].replace(':', ''): parameters[row,1:]})
+
+        averaging = int(info['Averaging'][1])
+        px_x = int(info['Pixel Area (X, Y, Z)'][1]) # 10
+        px_y = int(info['Pixel Area (X, Y, Z)'][2]) # 10
+        px_z = number_of_points = int(info['Pixel Area (X, Y, Z)'][3]) # 1024
+        description = info['Description'][1]
+
+        data_complete = []
+        final_metas = []
+        for y in range(0, px_y):
+            amplitude = gsf_a[y].reshape((1, px_x * px_z * averaging))[0]
+            phase = gsf_p[y].reshape((1, px_x * px_z * averaging))[0]
+            i = 0
+            f = i + px_z
+            for x in range(0, px_x):
+                for run in range(0, averaging):
+                    data_complete += [amplitude[i:f]]
+                    data_complete += [phase[i:f]]
+                    
+                    final_metas += [[run,x, y, self.channel_a]]
+                    final_metas += [[run,x, y, self.channel_p]]
+                    i = f
+                    f = i + px_z
+
+        return np.asarray(data_complete), info, final_metas
+
+    def _html_reader(self, path):
+        from pandas import read_html
+        try:
+            parameters = np.asarray(read_html(path)[0])
+        except ImportError:
+            print('''Probably you need to install xlml\nTry\n "pip install lxml"\nor\n"conda install -c conda-forge lxml"''')
+            pass
+
+        return parameters
+
+    def _gsf_reader(self, path):
```

Comment:
Please, never copy the same code to two places. We would need to have one function that both this reader and GSF reader would use. If the GSF does not allow reuse (code is not structured as such), we would need to change the GSF reader's code and then reuse it here.

Reply:
I fix this, but I used to return a little different.
So I created a global variable to change the flow, this global can be useful for more cases.

Reply:
Please do not do this. Using a global variable here makes the code prone to bugs. You should avoid using such programming structures at all costs for now. There are very few cases where global would be meaningful.

What you should do is to copy the contests of that function into a new one, which you would call from both GSF reader and you nea reader. For example, you could make a new function (and put it out of a class) from almost all of `read_spectra`, until `X = X.reshape((meta["YRes"], meta["XRes"]) + (1,))`. The input to that function would be a filename, and it would return X, XRr and YRr. Therefore, this new function would be usable in GSF, where `read_spectra` would now only have a few lines, and in your NEA reader.